pages/blog/2025-05-21-customizable-dashboards.mdx

```diff
@@ -0,0 +1,138 @@
+/**
+ * We want to build a ClickHouse query based on the query provided and the viewDeclaration that was selected.
+ * The final query should always follow this pattern:
+ * ```
+ *   SELECT
+ *     <...dimensions>,
+ *     <...metrics.map(metric => `${metric.aggregation}(${metric.alias})`>
+ *   FROM (
+ *      SELECT
+ *        <baseCte>.project_id,
+ *        <baseCte>.id
+ *        <...dimensions.map(dimension => `any(${dimension.sql}) as ${dimension.alias}`>,
+ *        <...metrics.map(metric => `${metric.sql} as ${metric.alias || metric.sql}`>
+ *      FROM <baseCte>
+ *      (...tableRelations.joinConditionSql)
+ *      WHERE <...filters>
+ *      GROUP BY <baseCte>.project_id, <baseCte>.id
+ *   )
+ *   GROUP BY <...dimensions>
+ *   ORDER BY <fields with directions>
+ * ```
+ */
+````
+
+To test this, we created our own query playground, which looked like the typical Langfuse playground, where you could input a JSON statement and it would compile and execute a corresponding query. Using these features, we started to replace the queries we used for our legacy/static home screen dashboard to verify that the query builder was sufficiently flexible and performant.
+
+We saw that the queries would execute at scale with a lot of flexibility. Given that we had already covered a lot of ground within those dashboards, seeing them run well and scale effectively even across longer time horizons gave us the confidence that we were on the right track and could proceed to build the remaining features.
+
+At this point, the power of ClickHouse came in really handy because initially, we optimized for development speed, focusing on understanding what is going on within the query builder. We had a previous one running on Postgres that covered a lot of edge cases and grew to more than 1,000 lines of code handling special cases. With the new query builder, we wanted to start with a small core and expand it once we saw the need. For most queries, the native case worked perfectly fine, which is why we stuck to it and decided to proceed with building and getting the feature into the hands of our users to gather feedback and develop in the direction that provides the most value to customers.
+
+In addition, we had to decide whether we wanted to embed all the widgets within the dashboards, meaning that they only live and exist within the scope of a single dashboard, or if we wanted to allow some reuse between the widgets. We found that the reuse option seemed very tempting, as it allows you to have a single source of truth on charts. The one expert who created a score can create a corresponding score widget, which can then be reused across multiple dashboards. I have seen in the past, with many companies, that the replication of things across dashboards either led to a lack of maintenance where items would get out of date or where a misunderstanding of the metric created multiple widgets with separate meaning.
+
+## How We Built It
```

Comment:
architecture mermaid would be cool, what are the core components, maybe even link to them in our code base or use filetree if applicable

Reply:
include all core abstractions in the chart + the public api as one interface